arangod/Aql/OptimizerRules.cpp

```diff
@@ -2944,11 +2945,90 @@ struct SortToIndexNode final : public WalkerWorker<ExecutionNode> {
   SortNode* _sortNode;
   std::vector<std::pair<Variable const*, bool>> _sorts;
   std::unordered_map<VariableId, AstNode const*> _variableDefinitions;
+  std::vector<std::vector<RegisterId>> _filters;
   bool _modified;
-
+ 
  public:
   explicit SortToIndexNode(ExecutionPlan* plan)
-      : _plan(plan), _sortNode(nullptr), _modified(false) {}
+      : _plan(plan), _sortNode(nullptr), _modified(false) {
+    _filters.emplace_back();
+  }
+ 
+  /// @brief gets the attributes from the filter conditions that can be proven
+  /// to be != null
+  void getNonNullAttributes(AstNode const* node, 
+                            Variable const* variable, 
+                            std::vector<std::vector<arangodb::basics::AttributeName>>& constAttributes,
+                            arangodb::HashSet<std::vector<arangodb::basics::AttributeName>>& nonNullAttributes) const {
+    std::pair<Variable const*, std::vector<arangodb::basics::AttributeName>> result;
+
+    if (node->type == NODE_TYPE_OPERATOR_BINARY_NE || 
+        node->type == NODE_TYPE_OPERATOR_BINARY_GT || 
+        node->type == NODE_TYPE_OPERATOR_BINARY_LT) {
+      AstNode const* lhs = node->getMember(0);
+      AstNode const* rhs = node->getMember(1);
+      AstNode const* check = nullptr;
+
+      if (lhs->isConstant() && 
+          lhs->isNullValue() && 
+          rhs->type == NODE_TYPE_ATTRIBUTE_ACCESS &&
+          node->type != NODE_TYPE_OPERATOR_BINARY_GT) {
+        // null != doc.value
+        // null < doc.value
+        check = rhs;
+      } else if (rhs->isConstant() && 
+                 rhs->isNullValue() && 
+                 lhs->type == NODE_TYPE_ATTRIBUTE_ACCESS && 
+                 node->type != NODE_TYPE_OPERATOR_BINARY_LT) {
+        // doc.value != null
+        // doc.value > null
+        check = lhs;
+      }
+      if (check != nullptr &&
+          check->isAttributeAccessForVariable(result, false) && 
+          result.first == variable) {
+        nonNullAttributes.emplace(std::move(result.second));
+      }
+    } else if (node->type == NODE_TYPE_OPERATOR_BINARY_EQ) {
+      AstNode const* lhs = node->getMember(0);
+      AstNode const* rhs = node->getMember(1);
+      AstNode const* check = nullptr;
+
+      if (lhs->isConstant() && rhs->type == NODE_TYPE_ATTRIBUTE_ACCESS) {
```

Comment:
```suggestion
      if (!lhs->isConstant() && rhs->type == NODE_TYPE_ATTRIBUTE_ACCESS) {
```

Reply:
does isConstant imply it is not null ? Otherwise wouldn't it match `FILTER null == doc.value` ?